src/switchroot/ostree-remount.c

```diff
@@ -54,20 +59,33 @@ do_remount (const char *target)
   struct statvfs stvfsbuf;
   if (statvfs (target, &stvfsbuf) == -1)
     return;
-  /* If no read-only flag, skip it */
-  if ((stvfsbuf.f_flag & ST_RDONLY) == 0)
+
+  const bool currently_writable = ((stvfsbuf.f_flag & ST_RDONLY) == 0);
+  if (writable == currently_writable)
     return;
-  /* It's a mounted, read-only fs; remount it */
-  if (mount (target, target, NULL, MS_REMOUNT | MS_SILENT, NULL) < 0)
-    {
-      /* Also ignore EINVAL - if the target isn't a mountpoint
-       * already, then assume things are OK.
-       */
-      if (errno != EINVAL)
-        err (EXIT_FAILURE, "failed to remount %s", target);
-    }
-  else
-    printf ("Remounted: %s\n", target);
+
+  int mnt_flags = MS_REMOUNT | MS_SILENT;
+  if (!writable)
+    mnt_flags |= MS_RDONLY;
+  if (mount (target, target, NULL, mnt_flags, NULL) < 0)
+    err (EXIT_FAILURE, "failed to remount %s", target);
+
+  printf ("Remounted %s: %s\n", writable ? "rw" : "ro", target);
+}
+
+static bool
+sysroot_is_configured_ro (void)
+{
+  struct stat stbuf;
+  static const char config_path[] = "/ostree/repo/config";
+  if (stat (config_path, &stbuf) != 0)
+    return false;
+
+  g_autoptr(GKeyFile) keyfile = g_key_file_new ();
+  if (!g_key_file_load_from_file (keyfile, config_path, 0, NULL))
+    return false;
+
+  return g_key_file_get_boolean (keyfile, "sysroot", "readonly", NULL);
```

Comment:
Why do we need to introduce a new key here instead of just handling it based on `ro` being in `/etc/fstab`? IIUC, for FCOS at least, it'll still be `rw` at initrd time to avoid the machine-id mess, and then will be re-mounted `ro` by systemd right before `ostree-remount.service` (though we could still recheck & remount here as we do right now).

Might've missed discussions around this, though keeping the same API as on other traditional systems I think is worth a lot.

Reply:
As of now, we don't have an `/etc/fstab` in FCOS anymore.

The other thing is...we're trying to create a "middle ground" around "read-only".  Here the physical root *is* writable - it has `/etc` and `/var` by default.  Wouldn't you expect having `ro` in `/etc/fstab` would be truly read-only?

This is really a magical ostree-specific flag I think to opt-in to new behavior for systems builders, not really something admins should see/change directly.

Reply:
> The other thing is...we're trying to create a "middle ground" around "read-only". Here the physical root is writable - it has /etc and /var by default. Wouldn't you expect having ro in /etc/fstab would be truly read-only?

OK, I follow. Hmm, I think one confusion here is that "read-only root" can mean a lot of things. Particularly `/etc` could be interpreted as being covered (and I think it'd be really cool to support this too). I wonder if `sysroot.readonly` shouldn't be an enum instead. E.g. `none` (status quo), `base` (everything except `/var` and `/etc`) and `full` (just `/var` is read-write). WDYT? (Not suggesting to implement `full` here, just keeping the option open so we leverage the same knob).